File: Coding-Patterns/Coding-Patterns-Python/05_greedy/non_overlapping_intervals.py
# LeetCode 435 - Non-overlapping Intervals
# Difficulty: Medium
# Pattern: Greedy
#
# Problem: Given an array of intervals, return the minimum number of
# intervals to remove to make the rest non-overlapping.
#
# Approach: Sort by end time. Greedily keep intervals with the earliest
# end — this leaves maximum room for future intervals.
# Count intervals we must remove (those that overlap with the last kept).
#
# Time: O(n log n)  Space: O(1)

from typing import List

def erase_overlap_intervals(intervals: List[List[int]]) -> int:
    intervals.sort(key=lambda x: x[1])  # sort by end time
    removed = 0
    prev_end = float('-inf')

    for start, end in intervals:
        if start >= prev_end:
            prev_end = end  # keep this interval
        else:
            removed += 1    # overlaps — remove it

    return removed


# ---- Tests ----
if __name__ == "__main__":
    assert erase_overlap_intervals([[1,2],[2,3],[3,4],[1,3]]) == 1
    assert erase_overlap_intervals([[1,2],[1,2],[1,2]]) == 2
    assert erase_overlap_intervals([[1,2],[2,3]]) == 0
    print("All tests passed.")

File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/05_greedy/NonOverlappingIntervals.cs
// LeetCode 435 - Non-overlapping Intervals
// Difficulty: Medium
// Pattern: Greedy (sort by end time)
//
// Time: O(n log n)  Space: O(1)

namespace CodingPatterns.Greedy;

public class NonOverlappingIntervals
{
    public int EraseOverlapIntervals(int[][] intervals)
    {
        // Step 1: Sort intervals based on end times
        Array.Sort(intervals, (a, b) => a[1] - b[1]);
        int removed = 0, prevEnd = int.MinValue;

        // Step 2: Iterate through intervals checking for overlaps
        foreach (var interval in intervals)
        {
            // If the current interval starts after or at the end of the previous interval, update prevEnd
            if (interval[0] >= prevEnd) 
                prevEnd = interval[1];
            // If there is an overlap, increment the removed count
            else 
                removed++;
        }
        return removed;
    }

    public static void Run()
    {
        var sol = new NonOverlappingIntervals();
        int[][] intervals = new int[][] {
            new int[] {1, 2},
            new int[] {2, 3},
            new int[] {3, 4},
            new int[] {1, 3}
        };
        int result = sol.EraseOverlapIntervals(intervals);
        Console.WriteLine($"Minimum number of intervals to remove: {result}");
    }
}
